llama.cpp

```diff
@@ -2980,6 +3011,89 @@ static void llm_load_tensors(
                         }
                     }
                 } break;
+            case LLM_ARCH_STABLELM:
+                {
+                    model.tok_embeddings = ml.create_tensor(ctx, tn(LLM_TENSOR_TOKEN_EMBD, "weight"), {n_embd, n_vocab}, GGML_BACKEND_CPU);
+
+                    // output
+                    {
+                        ggml_backend_type backend_norm;
+                        ggml_backend_type backend_output;
+
+                        // Don't allow for offloading of more than 33 layers.
+                        // Offloading 34 layers causes model to respond with letter 'E'
+                        // Offloading 35 layers doesn't work because of missing cuda implementation for rope:
+                        // GGML_ASSERT: ggml-cuda.cu:6402: ne00 == n_dims && "ne00 != n_dims is not implemented for CUDA yet"
+                        if (n_gpu_layers > 33) {
+                            n_gpu_layers = 33;
+                        }
+
```

Comment:
It should be implemented:

https://github.com/ggerganov/llama.cpp/blob/ba231e8a6dd8ad82acfe0e4d492ff7cef6b3f0a1/ggml-cuda.cu#L6481-L6507

Reply:
> This arch looks like the same as LLaMA but with a few bias tensors, correct?

Yes, but the `rope` is from `gpt-neox`

Reply:
in a recent-ish discussion we observed that `gpt-neox rope` and `llama rope` are **the same**, except they are permuted. https://github.com/jzhang38/TinyLlama/issues/24#issuecomment-1737334573

but i am not entirely sure this really is the case. but it worked for tinyllama (we had a working gpt-neox version working before llama). see discussion.

Reply:
Forgot to mention, `StableLM` has a different `n_dim = rope_pct * (hidden_size // num_attention_heads) = 20`

> Position Embeddings: Rotary Position Embeddings ([Su et al., 2021](https://arxiv.org/abs/2104.09864)) applied to the first 25% of head embedding dimensions for improved throughput following [Black et al. (2022)](https://arxiv.org/pdf/2204.06745.pdf).

~ https://github.com/Stability-AI/StableLM